### PR TITLE
Sprint 49

### PR DESCRIPTION
**Pull request by jhbertra**
# Continuous Integration base branch for Sprint 49

## Rules of the branch

- The idea is to integrate changes as frequently as possible. Open Pull Requests against this branch only if you really need your changes to be reviewed thoroughly.
- Otherwise push your changes directly to this branch
- Pull frequently
- Push frequently
- When pulling, use `git pull --ff-only` to avoid merge commits. If that doesn't work, please use `git rebase origin/sprint-49` to rebase your unpushed commits on the latest upstream commits rather than `git pull`
- If you break the build, please fix it!

## Changelog

- Fixed runtime issue with Halogen Hooks in forms #84b455809

### CODE REVIEW

File: marlowe/src/Language/Marlowe/Client/History.hs [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+      $ TypedScriptTxOut txOut md
+
+
+-- | Extract transaction outputs with datum.
+txDatums :: ChainIndexTx         -- ^ The transaction.
+         -> [(TxOut, TxOutRef)]  -- ^ The outputs that have datum.
+txDatums citx =
+  case citx ^. citxOutputs of
+    ValidTx txOuts -> filter (isJust . txOutDatumHash . fst)
+                      $ (fst &&& (TxOutRef (citx ^. citxTxId) . snd))
```
Comment (nau) [inline marlowe/src/Language/Marlowe/Client/History.hs:299]: It's a stylish remark, but could we try to use less funny Haskell and more boringly verbose Haskell? I mean, it's going to be 2 lines longer but easier to understand by non-veterans. And you could save of Control.Arrow import :)

File: marlowe/src/Language/Marlowe/Client/History.hs [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+  case citx ^. citxCardanoTx of
+    Just (SomeTx (ShelleyTx ShelleyBasedEraAlonzo tx@Alonzo.ValidatedTx{}) _) ->
+      catMaybes
+        [
+          do
+            (Alonzo.Data dat, _) <- Alonzo.indexedRdmrs tx $ Alonzo.Spending txin
+            pure (TxOutRef{..}, Redeemer . dataToBuiltinData $ dat)
+        |
+          txin@(Cardano.TxIn txid txix) <- S.toList . Alonzo.inputs . Alonzo.body $ tx
+        , let txOutRefId = TxId . toBuiltin . BS.drop 2 . toStrictByteString . toCBOR $ txid
```
Comment (nau) [inline marlowe/src/Language/Marlowe/Client/History.hs:332]: Why `drop 2`? "0x"?

File: marlowe/src/Language/Marlowe/Client/History.hs [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            -> [ChainIndexTx]      -- ^ The transactions at the Marlowe validator and role validator addresses.
+            -> MarloweTxOutRef     -- ^ The Marlowe transaction to start from.
+            -> Maybe History       -- ^ The sequence of subsequent redemptions.
+historyFrom address citxs consumed =
+  let
+    consumed' = tyTxOutRefRef consumed
+  in
+    -- The next redemption must consume the input.
+    case (any ((consumed' ==) . txInRef) . S.toList . (^. citxInputs)) `filter` citxs of
+      -- Only one transaction can consume the output of the previous step.
```
Comment (nau) [inline marlowe/src/Language/Marlowe/Client/History.hs:154]: A bit hard to understand the line of code. Awesome comments, though.

**Comment by jhbertra** [comment]
Closing and reopening to get Slack notifications

File: marlowe/src/Language/Marlowe/Client/History.hs (hunk 1 above)
Comment (bwbush) [reply marlowe/src/Language/Marlowe/Client/History.hs:299]: Yes, I'll replace the arrows with something easier to read.

File: marlowe/src/Language/Marlowe/Client/History.hs (hunk 3 above)
Comment (bwbush) [reply marlowe/src/Language/Marlowe/Client/History.hs:154]: I'll expand to make clearer. (Sorry about the pointfree style.)

File: marlowe/src/Language/Marlowe/Client/History.hs (hunk 2 above)
Comment (bwbush) [reply marlowe/src/Language/Marlowe/Client/History.hs:332]: Actually, I think it is a CBOR prefix. This is how other functions in higher-level Cardano packages do the conversion, but I couldn't find an existing conversion function from the Alonzo ledger to Plutus. This was bothering me, too, so I'll dig deeper.

File: marlowe/src/Language/Marlowe/Client/History.hs (hunk 1 above)
Comment (bwbush) [reply marlowe/src/Language/Marlowe/Client/History.hs]: Fixed in cb1fba3cd5ae838b556925ab12230ed5b4b6bc35.

File: marlowe/src/Language/Marlowe/Client/History.hs (hunk 3 above)
Comment (bwbush) [reply marlowe/src/Language/Marlowe/Client/History.hs:159]: Fixed in cb1fba3cd5ae838b556925ab12230ed5b4b6bc35.

File: marlowe/src/Language/Marlowe/Client/History.hs (hunk 2 above)
Comment (bwbush) [reply marlowe/src/Language/Marlowe/Client/History.hs]: Recorded as https://input-output.atlassian.net/browse/SCP-3503.

**Pull request merged**